File: recipe-java-runtime/src/main/java/ca/derekcormier/recipe/KeyedIngredient.java
package ca.derekcormier.recipe;

import com.fasterxml.jackson.annotation.JsonInclude;

@JsonInclude(JsonInclude.Include.NON_NULL)
public abstract class KeyedIngredient extends Ingredient {
  private String key;

  public KeyedIngredient(String name, String domain) {
    super(name, domain);
  }

  protected KeyedIngredient(KeyedIngredient other) {
    super(other);
    this.key = other.key;
  }

  public KeyedIngredient keyed(String key) {
    this.key = key;
    return this;
  }

  public String getKey() {
    return key;
  }

  protected void setKey(String key) {
    this.key = key;
  }

  public boolean hasKey() {
    return null != key;
  }
}

File: recipe-ts-runtime/src/KeyedIngredient.ts
import { Ingredient } from "./Ingredient";

export class KeyedIngredient extends Ingredient {
    private key: string | null = null;

    public constructor(name: string, domain: string) {
        super(name, domain);
    }

    protected setKey(key: string | null) {
        this.key = key;
    }

    public getKey(): string | null {
        return this.key;
    }

    public keyed(key: string | null): KeyedIngredient {
        this.key = key;
        return this;
    }

    public toJSON() {
        const jsonObj = super.toJSON();
        if (this.key) {
            jsonObj[this.getIngredientType()].key = this.key;
        }

        return jsonObj;
    }
}